### ISSUE TITLE
Add the possibility to get only the jre

### ISSUE THREAD

**Issue by jfouret** [conda-forge/openjdk-feedstock#152]
### Comment:

Hi, 

The whole temurin build for jdk is about 400 MB.  The jre only version is about 50-60MB.

A lot of recipes use this as dependency just to run an already built jar file which ends with a lot of unnecessary data transfer and storage. This also affects a lot a container images widely used.

Would it be possible to setup this recipe so that when there is "-jre" at the end of the version we use the jre only url that is also available as release via the github used here (https://github.com/adoptium/temurin21-binaries/releases).

If you agree with that with but lack time I can try an make a PR.

Thanks,

**Comment by gdonval** [comment]
I think there is a bit too much hardcoded dependence on `openjdk` for that to work well.

I'd suggest providing a `*_jre` variant instead so that `conda install ... openjdk=*=*_jre` does the right thing.

**Comment by XhstormR** [comment]
Yes，it is hoped that two variant options, JDK and JRE, can be provided for development compilation or just running Jar usage scenarios.

**Comment by bollwyvl** [comment]
Indeed, an `openjdk-headless` without all the x11 would be lovely, if someone was feeling feisty.

**Comment by h-vetinari** [comment]
PRs for `openjdk-jre` (naming?) and/or `openjdk-headless` are welcome.

**Comment by jfouret** [comment]
`-jre` is more adapted here, `-headless` is about not including `x11` libs. 

I think it's best to create another recipe from scratch. This repo has a quite high level of complexity adapted to provide an environment to build java code. The idea of providing a `-jre` is about minimizing the dependencies and reducing size to simply run already compiled projects.

In addition testing is also different for running only.

Regarding x11 libs, I think we shall not include them by default since it depends on the compiled program used. Users shall add them to their recipe, or to their environment directly if needed.

I will try to submit something here: `conda-forge/staged-recipes/recipes/openjdk-jre` but it will be linux only at first since I know nothing about windows.

**Comment by bollwyvl** [comment]
Looking at other ecosystems, it seems the `jre` and `headless` features are fairly intertwined. I _don't_ think separate feedstocks make a lot of sense, as, in my experience, all of the versions in an installed env will have to match up very tightly.

Probably the first, most time- and resource-effective step if there's going to be a lot of trial-by-CI would be to move to `rattler-build`, perhaps already starting with a multi-output recipe that initially just ships a single `openjdk`. However, without knowing more, doing the multi-output build _might_ blocked by the lack of [`cache`](https://github.com/conda-forge/conda-smithy/issues/2057), but I guess that remains to be seen. This is certainly a separate issue, and could start before any further things happen... and may well reveal corner cases in the `rattler-build(-conda-compat)` stack which need fixing anyway.

arch makes [jdk-openjdk](https://archlinux.org/packages/extra/x86_64/jdk-openjdk/) and [jre-openjdk-headless](https://archlinux.org/packages/extra/x86_64/jre-openjdk-headless/) mutually exclusive, which I don't feel like we want to emulate.

debian's [openjdk-11-jdk](https://packages.debian.org/buster/openjdk-11-jdk)(1mb), with the "meat" down at the `openjdk-11-jdk-headless`(80mb) and `openjdk-11-jre-headless`(40mb), and the heavy x11/etc deps in the `openjdk-11-jre`(0.2mb).

```mermaid
graph LR
    openjdk-11-jdk  --> openjdk-11-jdk-headless 
    openjdk-11-jdk --> openjdk-11-jre
    openjdk-11-jdk-headless --> openjdk-11-jre-headless
    openjdk-11-jre  --> openjdk-11-jre-headless
    openjdk-11-jre --> libgl & libx11 & fonts-dejavu-extra
    openjdk-11-jre-headless --> libasound & libgif & libfontconfig1 & libfreetype6
```

[fedora](https://packages.fedoraproject.org/pkgs/java-21-openjdk/java-21-openjdk/fedora-rawhide.html#dependencies) is very similar.

**Comment by jfouret** [comment]
I was not implying creating 2 new feedstock, just one "openjdk-jre" and make it headless by default.

The separation I was speaking about is (i) `jre only and/or headless` vs (ii) `jdk`.

The default behavious in other conda packages with a headless seems to be a recipe with 3 outputs:

- openjdk-jre (a base without the dependencies)
- openjdk-jre-headless (require the base plus any cli-sepcific requirements)
- openjdk-jre-graphics (require the base plus any gui-specific requirements)

I found an example where the variants were removed to include only the base as headless (pybsm). 

However for the java rune time, it makes sense to keep the variants are both are important for different users.

In addition pre-build jre for all arch/os are already provided by eclipse temurin therefore I would not rebuild them.

**Comment by bollwyvl** [comment]
#185 (built on #184, which is otherwise ready to go) fumblingly explores splitting the heavy x11 dependencies ([not working](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1172913&view=logs&j=7b6f2c87-f3a7-5133-8d84-7c03a75d9dfc&t=a78ca0b7-b297-5b86-8926-08275e0a9cdf&l=3854) yet).

Well and out of scope there would be determining how to split `$PREIFX/jvm/{bin,lib}` into the JRE and JDK sides of the house. As `conda-forge` _does_ build from source whenever possible, but pragmatically _can't_ yet for a number of platforms, a next step would be figuring out how to do the correct split of `bin` and `lib` for both build approaches.

> Looking again at, for example, debian's [jre](https://debian.pkgs.org/sid/debian-main-amd64/openjdk-23-jre-headless_23.0.2+7-1_amd64.deb.html#howto) and [jdk](https://debian.pkgs.org/sid/debian-main-amd64/openjdk-23-jdk-headless_23.0.2+7-1_amd64.deb.html#howto), it's not immediately obvious how one might make that call (e.g. everything in lib)

Basically, we want someone's env with a transient dependency on `openjdk-jre-headless` to be seamlessly upgradeable to an working one with a full-fat `openjdk` without re-installing packages... and certainly never to get inscrutable conflicts between a newer package that only has the "lightweight" dependency on `openjdk-jre-headless` and an older package-of-interest that depends on `openjdk`

**Comment by jfouret** [comment]
I have made a PR to add tumurin-jre only with an headless and graphics variants to be added in another feedstock. If you agree with that we can close this maybe.

**Comment by jfouret** [comment]
@h-vetinari Is there a plan for this to work in the short term ? 

I did not look into all the details of what @bollwyvl did but it seems blocked ? 

@bollwyvl what do you think of this PR https://github.com/conda-forge/staged-recipes/pull/29085 ?

Otherwise what need to be done in your #185 ?

**Comment by h-vetinari** [comment]
> Is there a plan for this to work in the short term ?

The plan in FOSS is that people who want/need a certain feature are also the ones to drive its design and implementation, in a way that is long-term maintainable (the latter half being what reviewers are concerned with first and foremost)

> I did not look into all the details of what [@bollwyvl](https://github.com/bollwyvl) did but it seems blocked ?

I don't see why it should be blocked. The conversion to rattler is a thing of convenience, not necessity. Beyond that, I'd recommend to keep work focused on the smallest reasonable increments. For example, we could add the jre builds without figuring out what's necessary for headless builds.

**Comment by jfouret** [comment]
> The plan in FOSS is that people who want/need a certain feature are also the ones to drive its design and implementation, in a way that is long-term maintainable (the latter half being what reviewers are concerned with first and foremost)

I am sorry I maybe have misphrased this. I was looking on how to take back things from here and I found this : 

> Welp, looking at #185, I don't see how we're going to get #152 without `--experimental`/`cache`

_Originally posted by @bollwyvl in https://github.com/conda-forge/openjdk-feedstock/issues/184#issuecomment-2650893055_

Hence my question. 

---

Regarding the implementation, it seems that things have changed in how we can build jre only from source:
- https://openjdk.org/jeps/220
- https://github.com/adoptium/temurin-build/issues/356

---

From what I got there are different ways

## Strategy 1
Build like it's done all and try to select files needed for jre afterwards

## Strategy 2
Try to build with specific target but that may differs depending on version

##

I will experiment both strategies locally and post here the results.

---

Anyway: 

- add the `outputs` key in the `meta.yaml`
  - openjdk (as-is)
  - openjdk-jre (to be implemented)
- keep only one build.sh file and work with an environment variable.
- split test files in 2 test_jre / test_jdk and run only test_jre in openjdk-jre, and both in openjdk
- split activate scripts for each output name
- Regarding runtime libraries / test and activate scripts for openjdk-jre I will work from https://github.com/jfouret/staged-recipes/tree/temurin-jre

--- 

From which branch should I work ? rattler-build PR #184  ? or main ? I was thinking `main` to make it easy to be merged later.

---

Timeline: it will take me at least 3-5 weeks since I have few moments I can work on this.

**Comment by h-vetinari** [comment]
> I was looking on how to take back things from here and I found this :
> 
> > Welp, looking at [#185](https://github.com/conda-forge/openjdk-feedstock/pull/185), I don't see how we're going to get [#152](https://github.com/conda-forge/openjdk-feedstock/issues/152) without `--experimental`/`cache`

That's a limitation which is imposed by the (otherwise benign) desire to translate the recipe to the up-and-coming v1 recipe format, but since that limitation is getting in the way of an actual use-case, my suggestion is very simply to postpone the translation to rattler-build. And then there's no reason why the jre cannot be split off correctly here.

One key part is that - at least from a high-level view - the jdk should be a superset of the jre, and whatever ends up packaged in the jre should not be duplicated in the jdk. That's why there's a very tight relationship between the two, and why these should be built on the same feedstock. For example, if you take
```yaml
outputs:
  - name: openjdk-jre
    ...
  - name: openjdk
    requirements:
      host:
        - {{ pin_subpackage("openjdk-jre", exact=True) }}
      run:
        - {{ pin_subpackage("openjdk-jre", exact=True) }}
```
... conda-build's snapshotting mechanism (what was installed in host before/after I ran the build script) will not repackage the bits from the jre in the jdk (that's the `host:` portion), but still ensure those bits are present at runtime (`run:`).

**Comment by jfouret** [comment]
What if we think of openjdk-jre as a subset of openjdk as follows:

```
# All variable defs as before
{% set files_to_include_in_jre = load_set_data('recipe/files_to_include.yaml') %}

package:
  name: openjdk
  version: {{ version }}

source:
  ...

build:
  number: 2
  ...
requirements:
  build:
    # No changes

outputs:
  - name: openjdk
    ...  # All existing sections (tests, requirements, etc.) as before
    files:
      exclude:
        - # Put here jre-specific activation scripts
    requirements:
      host:
        - # Same as before
      run:
        - # Same as before
    tests: # same as before
      
  - name: openjdk-jre
    files:
      include:
      {% for f in files_to_include_in_jre %}
        - {{ f }}
      {% endfor %}
    requirements:
      host:
      ...  # Add JRE-specific requirements here
      run:
      ...  # Add JRE-specific requirements here
    test:
      ...  # Add JRE-specific tests here (Maybe include a compiled jar file to be executed)

about:
  ...

extra:
  ...
```

- Creating a file that defines all that need to be kept for jre only seems fairly maintainable !? 

- Of note, if we forgot something in `/lib` related to the included binaries, the conda-build systems will raise an error. 

- In fact the `host` and `run` requirements does not need to be different from openjdk/openjdk-jre (or slightly) but if we want to include another subset openjdk-jre-headless it's good to implements the notions of different host and run requirements.

- I would not include the /man dir as we do this to reduce/optimize space...

Below is the list of files that I would include in jre (to-be-refined but that is the idea).

```
/lib/modules
/lib/libdt_socket.so
/lib/librmi.so
/lib/liblcms.so
/lib/jspawnhelper
/lib/libjawt.so
/lib/jfr/profile.jfc
/lib/jfr/default.jfc
/lib/libj2pcsc.so
/lib/libj2gss.so
/lib/libextnet.so
/lib/libfreetype.so
/lib/libmlib_image.so
/lib/libj2pkcs11.so
/lib/libfontmanager.so
/lib/libjava.so
/lib/libjdwp.so
/lib/libawt_xawt.so
/lib/jexec
/lib/libverify.so
/lib/libawt.so
/lib/libsctp.so
/lib/libinstrument.so
/lib/psfont.properties.ja
/lib/jvm.cfg
/lib/libmanagement.so
/lib/libzip.so
/lib/libsyslookup.so
/lib/libnio.so
/lib/libprefs.so
/lib/libmanagement_ext.so
/lib/libjsound.so
/lib/security/blocked.certs
/lib/security/cacerts
/lib/security/default.policy
/lib/security/public_suffix_list.dat
/lib/libjavajpeg.so
/lib/libsplashscreen.so
/lib/tzdb.dat
/lib/server/classes.jsa
/lib/server/classes_nocoops.jsa
/lib/server/libjvm.so
/lib/server/libjsig.so
/lib/libjsig.so
/lib/libnet.so
/lib/libawt_headless.so
/lib/libmanagement_agent.so
/lib/classlist
/lib/psfontj2d.properties
/lib/jrt-fs.jar
/lib/libjimage.so
/lib/libjli.so
/lib/libjaas.so
/bin/jfr
/bin/java
/bin/keytool
/bin/jrunscript
/bin/rmiregistry
/bin/jwebserver
/NOTICE
/release
/legal/**
/conf/**
```

**Comment by jfouret** [comment]
@h-vetinari would you agree to a PR toward the solution I described above where we do not change anything for the build but we rather select the files for each sub-packages ?

**Comment by h-vetinari** [comment]
As long as it works (and is tested), the file-based split is OK for me.

**Comment by jfouret** [comment]
I abandoned this because I was not able to make it work. 
I think of this again because the only software that is not instantly installed on pixi is the one depending on openjdk...
Still a huge impact on memory/bandwidth and therefore climate
Are you certain it's not fine to accept the Adoptium solution at this point ?

**Comment by h-vetinari** [comment]
> I was not able to make it work.

I did not see any PR here. Do this in the open, then other people can help you, collaborate, or take it as a starting point at some later date.

**Comment by h-vetinari** [comment]
This issue is still useful for other people (c.f. other comments here); in the meantime, staging builds for v1 recipes are unblocked now.

**Comment by jfouret** [comment]
Sorry I wanted to unsubscribe only.
